ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,355 @@
+                      upnpQueryFuture.complete(statusInfo);
+                    }
+
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void failure(
+                        final ActionInvocation invocation,
+                        final UpnpResponse operation,
+                        final String msg) {
+                      upnpQueryFuture.completeExceptionally(new Exception(msg));
+                    }
+                  };
+              upnpService.getControlPoint().execute(callback);
+
+              return upnpQueryFuture;
+            });
+  }
+
+  /**
+   * Sends a UPnP request to the discovered IGD to request a port forward.
+   *
+   * @param portMapping is a portMapping object describing the desired port mapping parameters.
+   * @return A CompletableFuture that can be used to query the result (or error).
+   */
+  public CompletableFuture<String> requestPortForward(final PortMapping portMapping) {
+
+    CompletableFuture<String> upnpQueryFuture = new CompletableFuture<>();
+
+    return discoverService(SERVICE_TYPE_WAN_IP_CONNECTION)
+        .thenCompose(
+            service -> {
+
+              // our query, which will be handled asynchronously by the jupnp library
+              PortMappingAdd callback =
+                  new PortMappingAdd(service, portMapping) {
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void success(final ActionInvocation invocation) {
+                      // TODO: return value here?
+                      upnpQueryFuture.complete("TODO");
+                    }
+
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void failure(
+                        final ActionInvocation invocation,
+                        final UpnpResponse operation,
+                        final String msg) {
+                      upnpQueryFuture.completeExceptionally(new Exception(msg));
+                    }
+                  };
+              upnpService.getControlPoint().execute(callback);
+
+              return upnpQueryFuture;
+            });
+  }
+
+  /**
+   * Recursively crawls the given device to look for specific services.
+   *
+   * @param device is the device to inspect for desired services.
+   * @param serviceTypes is a set of service types to look for.
+   */
+  @SuppressWarnings("rawtypes")
+  protected void inspectDeviceRecursive(final Device device, final Set<String> serviceTypes) {
+    for (Service service : device.getServices()) {
+      String serviceType = service.getServiceType().getType();
+      if (serviceTypes.contains(serviceType)) {
+        // TODO: handle case where service is already "recognized" as this could lead to
+        // some odd bugs
+        recognizedServices.put(serviceType, service);
+        LOG.info("Discovered service " + serviceType);
```

Comment:
please log this at debug.  I don't think we want to enumerate everyones chromecast and kodi player at startup,  info is the default log level for the console.  Info is reserved for operational status and it would be the correct level to log "we mapped this port"

Reply:
Removed -- this was redundant with another log @ debug level. Yes, this would annoy most people at info :)